vsintegration/src/FSharp.Editor/QuickInfo/QuickInfoProvider.fs

```diff
@@ -53,45 +53,70 @@ type internal FSharpQuickInfoProvider
         let solution = workspace.CurrentSolution
 
-        let documentId (range: range) =
+        /// Retrieve the DocumentId from the workspace for the range's file
+        let docIdOfRange (range: range) =
             let filePath = System.IO.Path.GetFullPathSafe range.FileName
-            let projectOf (id : DocumentId) = solution.GetDocument(id).Project
 
-            //The same file may be present in many projects. We choose one from current or referenced project.
-            let rec matchingDoc = function
-            | [] -> None
-            | id::_ when projectOf id = thisDoc.Project || IsScript thisDoc.FilePath -> Some id
-            | id::tail -> 
-                if (projectOf id).GetDependentProjects() |> Seq.contains thisDoc.Project then Some id
-                else matchingDoc tail
-            solution.GetDocumentIdsWithFilePath(filePath) |> List.ofSeq |> matchingDoc
+            // The same file may be present in many projects. We choose one from current or referenced project.
+            let candidates = solution.GetDocumentIdsWithFilePath filePath
+            match candidates.Length with 
+            | 0 -> None 
+            | 1 -> 
+                let docId = candidates.[0]
+                if docId.ProjectId = initialDoc.Id.ProjectId || IsScript initialDoc.FilePath then Some docId else None
+            | _ -> 
+                candidates |> Seq.tryFind (fun docId -> 
+                    solution.GetDependentProjects docId.ProjectId |> Seq.contains initialDoc.Project
+                )
 
         let canGoTo range =
-            range <> rangeStartup && documentId range |> Option.isSome
-
-        let goTo range = 
-            asyncMaybe { 
-                let! id = documentId range
-                let doc = solution.GetDocument(id)
-                let! src = doc.GetTextAsync()
-                let! span = CommonRoslynHelpers.TryFSharpRangeToTextSpan(src, range)
-                let goToService = doc.Project.LanguageServices.GetService<IGoToDefinitionService>()
-                if not (isNull goToService) && goToService.TryGoToDefinition(doc, span.Start, Async.DefaultCancellationToken) then
-                   let! session = SessionHandling.currentSession
-                   session.Dismiss()   
-            } |> Async.Ignore |> Async.StartImmediate 
-
-        let formatMap = typemap.ClassificationFormatMapService.GetClassificationFormatMap("tooltip")
-
-        let props =
+            range <> rangeStartup && docIdOfRange range |> Option.isSome
+
+        // to ensure proper navigation decsions we need to check the type of document the navigation call
+        // is originating from and the target we're provided by default
+        //  - signature files (.fsi) should navigate to other signature files 
+        //  - implementation files (.fs) should navigate to other implementation files
+        //let adjustTarget (range:range) =
+
+        let navigateTo (range:range) = 
+            maybe { 
```

Comment:
This may be subjective but for me async version works better. There still are occasional slowdowns but now there is noticeable UI freeze.